File: 项目开发进度2009年2月9日/SeeMapCell/MapXtreme/6.6/Samples/WebControlsSourceCode/cs/LayerControl/LayerController.cs
using System;
using System.IO;
using System.Text.RegularExpressions;
using System.Web;
using System.Web.SessionState;
using System.Xml;
using System.Xml.XPath;
using System.Xml.Xsl;

namespace MapInfo.WebControls
{
	/// <summary>
	/// Handler for the LayerController.
	/// </summary>
	/// <remarks>The HTML for this control is generated on demand when a map changes because the user interface contains values that 
	/// need updating when the map changes.</remarks>
	public class LayerController : IHttpHandler, IRequiresSessionState
	{
		/// <summary>
		/// Reuses the instance of this handler.
		/// </summary>
		/// <value>Returns <c>true</c>.</value>
		/// <remarks>None</remarks>
		public bool IsReusable {
			get {
				return true;
			}
		}

		/// <summary>
		/// Processes requests for the LayerControl.
		/// </summary>
		/// <param name="context">HttpContext</param>
		/// <remarks>This method parses the URL. Gets the model for the LayerControl from the session and then uses that to get the XML. Then the HTML is generated
		/// by using XSLT transform and returnes back as response.
		/// </remarks>
		public void ProcessRequest(HttpContext context)
		{
			string _command = context.Request["Command"];
			string _uniqueID = context.Request["UniqueID"];
			string _mapAlias = context.Request["MapAlias"];
			string _mapControlID = context.Request["MapControlID"];
			// Get Xslt file name
			string _xsltFile = context.Request["XsltFile"];
			// Get the resource path name passed as data and properly encode it.
			byte[] buffer = new byte[context.Request.InputStream.Length];
			int read = context.Request.InputStream.Read(buffer, 0, (int)context.Request.InputStream.Length);
			string _resourcePath = System.Text.UTF8Encoding.UTF8.GetString( buffer );
			string _imgPath = _resourcePath;
			// Form the full path for xslt file
			string _xsltPath = context.Server.MapPath(_resourcePath) + "\\" + _xsltFile;
			LayerControlModel layerModel = LayerControlModel.SetDefaultModelInSession();

			switch(_command) {
				case "GetHTML":
					XmlDocument treeDoc = layerModel.GetLayerXML(_mapAlias,  _uniqueID, _imgPath);

					// load the xslt file into reader
					System.IO.TextReader tr = new System.IO.StreamReader(_xsltPath);
					MemoryStream ms = new MemoryStream();
					StreamWriter sw = new StreamWriter(ms);
					// Now replace the tags in the xslt file with actual strings from resources
					string line = "";
					while((line = tr.ReadLine()) != null) {
						Regex regexp = new Regex("(Treeview2_([0-9]+))");
						MatchCollection mc = regexp.Matches(line);
						for (int i=mc.Count-1; i>=0;i--) {
							Match m = mc[i];
							Group g = m.Groups[0];
							Capture c = m.Groups[0].Captures[0];
							string resName = line.Substring(c.Index, c.Length);
							if (resName.Equals("Treeview2_3")) {
								line = line.Replace(resName, Resources.ResourceFolder);
							} else {
								string resString = L10NUtils.Resources.GetString(resName);
								line = line.Replace(resName, resString);
							}
						}
						sw.WriteLine(line);
					}

					sw.Flush();
					ms.Position = 0;

					// Use that TextReader as the Source for the XmlTextReader
					System.Xml.XmlReader xr = new System.Xml.XmlTextReader(ms);
					// Create a new XslTransform class
					System.Xml.Xsl.XslTransform treeView = new System.Xml.Xsl.XslTransform();
					// Load the XmlReader StyleSheet into the XslTransform class
					treeView.Load(xr, null, null);

					StringWriter sw2 = new StringWriter();

					XPathNavigator nav = treeDoc.CreateNavigator();

					// Do the transform and write to response
					treeView.Transform(nav, null,  sw2,  null);
					context.Response.Write(sw2.ToString());
					tr.Close();
					sw.Close();
					ms.Close();
					sw2.Close();
					
					/*
					XmlDocument treeDoc = layerModel.GetLayerXML(_mapAlias,  _uniqueID, _imgPath);

					XslTransform treeView = new XslTransform();
					treeView.Load(_xsltPath);
   
					StringWriter sw = new StringWriter();

					XPathNavigator nav = treeDoc.CreateNavigator();

					// Do the transform and write to response
					treeView.Transform(nav, null,  sw,  null);
					context.Response.Write(sw.ToString());
*/					
					break;
			}
		}
	}
}

File: MapXtreme/6.6/Samples/WebControlsSourceCode/vb/LayerControl/LayerController.vb
Imports System
Imports System.IO
Imports System.Web
Imports System.Web.SessionState
Imports System.Xml
Imports System.Xml.XPath
Imports System.Xml.Xsl
Imports System.Text.RegularExpressions


Public Class LayerController
    Implements IHttpHandler, IRequiresSessionState

    Public ReadOnly Property IsReusable() As Boolean Implements IHttpHandler.IsReusable
        Get
            Return False
        End Get
    End Property

    Public Sub ProcessRequest(ByVal context As HttpContext) Implements IHttpHandler.ProcessRequest
        Dim command As String = context.Request("Command")
        Dim uniqueID As String = context.Request("UniqueID")
        Dim mapAlias As String = context.Request("MapAlias")
        Dim mapControlID As String = context.Request("MapControlID")
        ' Get Xslt file name
        Dim xsltFile As String = context.Request("XsltFile")
        ' Get the resource path name passed as data and properly encode it.
        Dim buffer() As Byte = New Byte(context.Request.InputStream.Length) {}
        Dim read As Integer = context.Request.InputStream.Read(buffer, 0, CType(context.Request.InputStream.Length, Integer))
        Dim resourcePath As String = System.Text.UTF8Encoding.UTF8.GetString(buffer, 0, read)
        ' in VB trim the null out of the string. CS works fine without trimming.
        Dim imgPath As String = resourcePath
        ' Form the full path for xslt file
        Dim xsltPath As String = context.Server.MapPath(resourcePath) + "\\" + xsltFile

        Dim layerModel As LayerControlModel = LayerControlModel.SetDefaultModelInSession()

        Select Case command
            Case "GetHTML"

                Dim treeDoc As XmlDocument = layerModel.GetLayerXML(mapAlias, uniqueID, imgPath)

                ' load the xslt file into reader
                Dim tr As System.IO.TextReader = New System.IO.StreamReader(xsltPath)
                Dim ms As MemoryStream = New MemoryStream
                Dim sw As StreamWriter = New StreamWriter(ms)
                ' Now replace the tags in the xslt file with actual strings from resources
                Dim line As String = ""
                While True
                    line = tr.ReadLine()
                    If (line Is Nothing) Then
                        Exit While
                    End If
                    Dim regexp As Regex = New Regex("(Treeview2_([0-9]+))")
                    Dim mc As MatchCollection = regexp.Matches(line)
                    Dim i As Integer
                    For i = mc.Count - 1 To 0 Step i - 1
                        Dim m As Match = mc(i)
                        Dim g As Group = m.Groups(0)
                        Dim c As Capture = m.Groups(0).Captures(0)
                        Dim resName As String = line.Substring(c.Index, c.Length)
                        If (resName.Equals("Treeview2_3")) Then
                            line = line.Replace(resName, Resources.ResourceFolder)
                        Else
                            Dim resString As String = L10NUtils.Resources.GetString(resName)
                            line = line.Replace(resName, resString)
                        End If
                    Next
                    sw.WriteLine(line)
                End While

                sw.Flush()
                ms.Position = 0

                ' Use that TextReader as the Source for the XmlTextReader
                Dim xr As System.Xml.XmlReader = New System.Xml.XmlTextReader(ms)
                ' Create a new XslTransform class
                Dim treeView As System.Xml.Xsl.XslTransform = New System.Xml.Xsl.XslTransform
                ' Load the XmlReader StyleSheet into the XslTransform class
                treeView.Load(xr, Nothing, Nothing)

                Dim sw2 As StringWriter = New StringWriter

                Dim nav As XPathNavigator = treeDoc.CreateNavigator()

                ' Do the transform and write to response
                treeView.Transform(nav, Nothing, sw2, Nothing)
                context.Response.Write(sw2.ToString())
                tr.Close()
                sw.Close()
                ms.Close()
                sw2.Close()
                Return
        End Select
    End Sub

End Class
